rustls/src/msgs/codec.rs

```diff
@@ -174,88 +177,149 @@ impl Codec for u64 {
     }
 }
 
-pub fn encode_vec_u8<T: Codec>(bytes: &mut Vec<u8>, items: &[T]) {
-    let len_offset = bytes.len();
-    bytes.push(0);
+/// Represent a Rust `Vec` together with the length prefix type used to encode it in TLS.
+///
+/// This implements various traits to make it easy to use in place of a normal `Vec` and to
+/// easily convert between `Vec` and `TlsVec` at API boundaries.
+#[derive(Clone, Debug)]
+pub struct TlsVec<L, T> {
+    items: Vec<T>,
+    len: PhantomData<L>,
+}
+
+impl<L: LengthPrefix, T: Codec> Codec for TlsVec<L, T>
+where
+    usize: From<L>,
+{
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        let len_offset = bytes.len();
+        bytes.extend(L::ZERO_LEN);
+
+        for i in &self.items {
+            i.encode(bytes);
+        }
 
-    for i in items {
-        i.encode(bytes);
+        let len = bytes.len() - len_offset - L::BYTES;
+        L::write_into(len, len_offset, bytes);
```

Comment:
This might fail if `len` is larger than what can be encoded in `L::BYTES` bytes, right?

There is no need for `L::write_into` to be implemented for each type; it is easy to write a non-generic function that does it.